Search unit .d/ drop-in files also in all "dash prefixes" of a unit

poettering opened the pull request:
With this PR for a unit `foo-bar-baz.service` we'll all three of the following for unit files:

1. `foo-bar-baz.service.d/` (as before)
2. `foo-bar-.service.d/`
3. `foo-.service.d/`

This was discussed in #7599 before.

bl33pbl0p commented:
Would applying global defaults be then supported with -.service.d?

poettering commented:
> Would applying global defaults be then supported with -.service.d?

For now I decided not to do add support for such a "root drop-in". Maybe we can add that later on, but for now I wanted to stay away from that, it smells like a very problematic concept to me since it would allow changing defaults for very much unrelated services — which is something I fear might trigger tons of issues.

bl33pbl0p commented:
I realised that too. For example, you really don't override ExecStart= for every service but if you use -.service.d to do that, it would work and that does not make any sense. If this would be added, it would be better to probably put it in system.conf under various sections apart from [Manager] (but I figure that config file is strictly for the manager, and including too many things in it would make it really big, so it should be probably be somewhere else like service.conf), and only include commented out settings that are allowed to be set globally, or some similar mechanism (OnFailure= makes a lot of sense to be set globally, while something like Before= or After= does not make much sense).

Thanks for this anyway, very useful feature.

bl33pbl0p commented:
I think this could be documented how these changes happen from a higher-level path to a lower-level path. It seems logically correct the path at the highest level takes the most preference and the service has the least preference, when keys are being overriden. Therefore, one could symlink 01-foo.conf to /dev/null in `foo-bar-baz.service.d` to mask changes made by foo-bar-.service.d/01-foo.conf. Other than that, I tested that and it works well!

poettering commented:
Hmm, so I think this needs to be reworked. It currently imports the more specific snippets first, and the more generic snippets last. This really should be reversed I figure, so that stuff down the tree matters most, and up the tree matters least.

Will rework and then add docs for it.

poettering commented:
Force pushed a new version now that drops some unrelated patches, and documents the overriding order, and some other minor changes.

PTAL!

poettering commented:
(oh, and I also added a test for this, the verifies the overriding order)

poettering commented:
Force pushed another version, that hopefully fixes the CI failures

keszybz commented on src/basic/unit-name.c:
```diff
@@ -211,6 +211,26 @@ int unit_name_to_prefix_and_instance(const char *n, char **ret) {
+
+        s = strdup(e);
+        if (!s)
+                return -ENOMEM;
```
Hmm, wouldn't we be better off returning a static string? Something like
```C
assert_se(s = unit_type_to_string(unit_type_from_string(e)));
return s;
```
This doesn't have the dot, but it can be easily added by the caller.

poettering replied:
hmm, so, this is actually pretty bad... currently we have the unit "type" which is without the dot, and the unit "suffix" which includes the dot. The "type" is an enum named `UnitType` plus the usual from_string() + to_string() calls, hence I think it's pretty natural that they come without dots. Now, the suffix is otoh is used whenver we split or put together unit names, most prominently in unit_name_build(), where I think it makes sense to include it...

Grmbl... I see if I find a nice solution for this

keszybz commented on src/shared/dropin.c:
```diff
@@ -128,14 +128,12 @@ static int unit_file_find_dir(
+        if (r < 0) /* Also, downgrade logging about -ENAMETOOLONG to LOG_DEBUG, users are not even able to create the
+                    * drop-in dir in such case. This mostly happens for device units with long /sys path. */
+                return log_full_errno(r == -ENAMETOOLONG ? LOG_DEBUG : LOG_WARNING, r,
+                                      "Failed to canonicalize path %s: %m", path);
```
That changes the behaviour, because an error is returned, which causes unit_file_find_dirs() to abort early.

poettering replied:
indeed

poettering commented:
OK, force pushed a new version. Two changes:

1. I dropped unit_name_to_suffix() again, and instead added unit_name_build_from_type. This is a bit a nicer I think, as we don't bother with passing strings around anymore, but instead just use the enum. Given that we always check the strings against the list of enums this is more efficient that way…

2. I fixed the ENAMETOOLONG thing.

PTAL

keszybz commented on src/basic/unit-name.c:
```diff
@@ -241,25 +241,45 @@ int unit_name_change_suffix(const char *n, const char *suffix, char **ret) {
+        if (isempty(suffix))
+                return -EINVAL;
+        if (suffix[0] != '.')
+                return -EINVAL;
```
The first check is actually a subset of the second...

And shouldn't this be an `assert` instead?

poettering replied:
true, the first check is redundant indeed...

I think it's a good idea to place assert()s to catch programming errors while catching invalid data errors with runtime errors. I'd claim that a parameter being NULL where it shouldn't is really a programming error, while looking into a string could be an input data error, and hence should be handled more defensively...

keszybz replied:
I prepped a short patch for this locally. Let's discuss it in the new PR.

keszybz commented:
`E: Could not get lock /var/lib/dpkg/lock - open (11: Resource temporarily unavailable)` — we have seen this one before.
